### DISCUSSION TITLE
Perception TF v1 code savedModel graph says it has no operation

### DISCUSSION THREAD

**Discussion by Litchilitchy** [rasbt/deeplearning-models#53] [Q&A]
hi I am using this repo code
```
g = tf.Graph()
with g.as_default():
    # Graph Inputs
    features = tf.placeholder(dtype=tf.float32,
                              shape=[None, 2], name='features')
    targets = tf.placeholder(dtype=tf.float32,
                             shape=[None, 1], name='targets')

    # Model Parameters
    weights = tf.Variable(tf.zeros(shape=[2, 1],
                                   dtype=tf.float32), name='weights')
    bias = tf.Variable([[0.]], dtype=tf.float32, name='bias')

    # Forward Pass
    linear = tf.add(tf.matmul(features, weights), bias, name='linear')
    ones = tf.ones(shape=tf.shape(linear))
    zeros = tf.zeros(shape=tf.shape(linear))
    prediction = tf.where(condition=tf.less(linear, 0.),
                          x=zeros,
                          y=ones,
                          name='prediction')

    # Backward Pass
    errors = targets - prediction
    weight_update = tf.assign_add(weights,
                                  tf.reshape(errors * features, (2, 1)),
                                  name='weight_update')
    bias_update = tf.assign_add(bias, errors,
                                name='bias_update')

    train = tf.group(weight_update, bias_update, name='train')

    saver = tf.train.Saver(name='saver')
```
and save it using
```
inputs = dict([(features.name, features)])
outputs = dict([(prediction.name, prediction)])
tf.saved_model.simple_save(sess, "my_path", inputs, outputs)
```

and I can use `saved_model_cli` to see the model, following is part of it
```
signature_def['serving_default']:
  The given SavedModel SignatureDef contains the following input(s):
    inputs['features:0'] tensor_info:
        dtype: DT_FLOAT
        shape: (-1, 2)
        name: features:0
```
but when I use TF2 `tf.keras.model.load_model("my_path")` it raise error `KeyError: "The name 'features:0' refers to a Tensor which does not exist. The operation, 'features', does not exist in the graph."`, using java api raise the similar error.

Could this save to savedModel? How should I do it correctly?

**Comment by rasbt** [comment]
oh this is Tf v1 code, I don't think this is readily compatible with TensorFlow 2

**Reply by Litchilitchy** [reply]
it is the model problem, I use TF 1.15 java api the error is the same

**Reply by Litchilitchy** [reply]
issue solved, the save operation should be in `with Session`

**Reply by rasbt** [reply]
Ok great it's solved :). Was it something that was wrong on your end or in the notebook (just asking so I can fix it).

**Reply by Litchilitchy** [reply]
nothing wrong, just need to code in `with Session` if export savedModel. However, if code outside this block it could also save with no exception but could not be used, so weird.

**Comment by NCEPU-zny** [comment]
view it on GitHub



At 2021-03-29 10:15:25, "Song Jiaming" ***@***.***> wrote:

nothing wrong, just need to code in with Session if export savedModel. However, if code outside this block it could also save with no exception but could not be used, so weird.
